pkg/observable/channel/observable.go

```diff
@@ -0,0 +1,148 @@
+	}
+
+	// start listening to the producer and emit values to observers
+	go obs.goListen(obs.producer)
+
+	return obs, obs.producer
+}
+
+// WithProducer returns an option function sets the given producer in an observable
+// when passed to NewObservable().
+func WithProducer[V any](producer chan V) option[V] {
+	return func(obs *channelObservable[V]) {
+		obs.producer = producer
+	}
+}
+
+// Subscribe returns an observer which is notified when the producer channel
+// receives a value.
+func (obsvbl *channelObservable[V]) Subscribe(ctx context.Context) observable.Observer[V] {
+	obsvbl.observersMu.Lock()
+	defer obsvbl.observersMu.Unlock()
+
+	observer := NewObserver[V](ctx, obsvbl.onUnsubscribeFactory)
+	obsvbl.observers = append(obsvbl.observers, observer)
+
+	// caller can rely on context cancellation or call Close() to unsubscribe
+	// active observers
+	if ctx != nil {
+		// asynchronously wait for the context to close and unsubscribe
+		go goUnsubscribeOnDone[V](ctx, observer)
+	}
+	return observer
+}
+
+func (obsvbl *channelObservable[V]) Close() {
+	obsvbl.close()
+}
+
+// CONSIDERATION: decide whether this should close the producer channel; perhaps
+// only if it was provided.
+func (obsvbl *channelObservable[V]) close() {
+	obsvbl.observersMu.Lock()
+	defer obsvbl.observersMu.Unlock()
+
+	observers := obsvbl.observers
+
+	for _, obsvr := range observers {
+		fmt.Printf("channelObservable#goListen: unsubscribing %p\n", obsvr)
+		obsvr.Unsubscribe()
+	}
+
+	// clear observers
+	obsvbl.observers = []*channelObserver[V]{}
+}
+
+// goListen to the producer and notify observers when values are received. This
+// function is blocking and should be run in a goroutine.
+func (obsvbl *channelObservable[V]) goListen(producer <-chan V) {
+	for notification := range producer {
+		// copy observers to avoid holding the lock while notifying
+		obsvbl.observersMu.RLock()
+		observers := make([]*channelObserver[V], len(obsvbl.observers))
+		for i, obsvr := range obsvbl.observers {
+			observers[i] = obsvr
+		}
+		obsvbl.observersMu.RUnlock()
+
+		// notify observers
+		for _, obsvr := range observers {
+			// CONSIDERATION: perhaps continue trying to avoid making this
+			// notification async as it would effectively use goroutines
+			// in memory as a buffer (with little control surface).
+			obsvr.notify(notification)
+		}
+	}
+
+	// Here we know that the producer has been closed, all observers should be closed as well
+	obsvbl.close()
+}
+
+// goUnsubscribeOnDone unsubscribes from the subscription when the context is.
+// It is blocking and intended to be called in a goroutine.
+func goUnsubscribeOnDone[V any](ctx context.Context, subscription observable.Observer[V]) {
+	<-ctx.Done()
+	subscription.Unsubscribe()
+}
+
+// onUnsubscribeFactory returns a function that removes a given channelObserver from the
+// observable's list of observers.
+func (obsvbl *channelObservable[V]) onUnsubscribeFactory() UnsubscribeFunc[V] {
```

Comment:
Just now noticing that this doesn't need to be a factory, it can just be the function that it currently returns.